File: ch4/clip7/boj2624/main.py
t = int(input())
k = int(input())
d = [0] * (t + 1)
d[0] = 1

for i in range(k):
    coin_price, coin_count = map(int, input().split())
    for j in range(t, -1, -1):
        for cnt in range(1, coin_count + 1):
            next_value = j + coin_price * cnt
            if next_value > t:
                break
            d[next_value] += d[j]

print(d[t])

File: ch4/clip7/boj2624/Main.java
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        int k = sc.nextInt();
//        int dp[][] = new int[k + 1][t + 1];
//
//        dp[0][0] = 1;
//        for (int i = 1; i <= k; i++) {
//            int coinPrice = sc.nextInt();
//            int coinCount = sc.nextInt();
//            for (int value = 0; value <= t; value++) {
//                for (int cnt = 0; cnt <= coinCount; cnt++) {
//                    int nextValue = value + coinPrice * cnt;
//                    if (nextValue > t) break;
//                    dp[i][nextValue] += dp[i - 1][value];
//                }
//            }
//        }
//        System.out.println(dp[k][t]);
        int[] dp = new int[t + 1];
        dp[0] = 1;
        for (int i = 0; i < k; i++) {
            int coinPrice = sc.nextInt();
            int coinCount = sc.nextInt();
            for (int j = t; j >= 0; j--) {
                for (int cnt = 1; cnt <= coinCount; cnt++) {
                    int nextValue = j + coinPrice * cnt;
                    if (nextValue > t) break;
                    dp[nextValue] += dp[j];
                }
            }
        }
        System.out.println(dp[t]);
    }
}
